### PR TITLE
tests/hiop: convert to new stand-alone test process

### PR DESCRIPTION
**Pull request by tldahlgren**
UPDATE: Converted the stand-alone tests to the new process.

```
$ spack -v test run hiop
==> Spack test nt6phaz534b4srmsc7qdtvtcigwmfzuj
==> Testing package hiop-0.7.2-yjkmatz
==> [2023-06-13-17:06:04.083444] test: test_nlpmdsex1: run NlpMDsEx1.exe selfchecks
==> [2023-06-13-17:06:04.083680] test: test_nlpmdsex1_0: run example selfcheck
==> [2023-06-13-17:06:04.085131] 'SPACK_ROOT/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/hiop-0.7.2-yjkmatzkgrzjhqjzlw4e62tqpp3xds2t/bin/NlpMdsEx1.exe' '400' '100' '0' '-selfcheck'
===============
Hiop SOLVER
===============
Using 1 MPI ranks.
---------------
Problem Summary
---------------
Total number of variables: 900
     lower/upper/lower_and_upper bounds: 401 / 401 / 1
Total number of equality constraints: 400
Total number of inequality constraints: 3
     lower/upper/lower_and_upper bounds: 2 / 2 / 1
iter    objective     inf_pr     inf_du   lg(mu)  alpha_du   alpha_pr linesrch
   0  3.9800005e+02 4.990e+02  4.000e+00  -1.00  0.000e+00  0.000e+00  -(-)
   1  1.8397701e+01 1.818e+02  1.457e+00  -1.00  3.341e-01  6.357e-01  1(s)
   2 -4.3390041e+01 5.305e+01  4.253e-01  -1.00  7.654e-01  7.081e-01  1(s)
   3 -4.9598500e+01 8.260e-13  2.539e-01  -1.00  8.152e-01  1.000e+00  1(s)
   4 -4.9406538e+01 7.556e-13  3.078e+01  -1.00  9.388e-01  5.909e-02  1(f)
   5 -4.3566407e+01 8.954e-13  1.000e-06  -1.00  1.000e+00  1.000e+00  1(f)
   6 -4.9711482e+01 1.172e-12  1.601e+00  -2.55  9.261e-01  1.000e+00  1(f)
   7 -4.9974904e+01 1.388e-13  2.828e-08  -2.55  1.000e+00  1.000e+00  1(f)
   8 -4.9992701e+01 8.282e-14  1.504e-09  -3.82  1.000e+00  1.000e+00  1(f)
   9 -4.9993494e+01 2.287e-14  1.504e-09  -3.82  1.000e+00  1.000e+00  1(f)
iter    objective     inf_pr     inf_du   lg(mu)  alpha_du   alpha_pr linesrch
  10 -4.9993744e+01 1.221e-14  1.732e-03  -5.73  9.710e-01  1.000e+00  1(f)
  11 -4.9994740e+01 1.342e-12  1.845e-11  -5.73  1.000e+00  1.000e+00  1(f)
  12 -4.9994726e+01 1.354e-14  1.845e-11  -5.73  1.000e+00  1.000e+00  1(f)
  13 -4.9994868e+01 1.110e-14  3.029e-04  -6.00  8.509e-01  1.000e+00  1(f)
  14 -4.9994888e+01 7.421e-13  1.000e-11  -6.00  1.000e+00  1.000e+00  1(f)
Successfull termination.
Total time 0.148s  
Hiop internal time:     total 0.082s      avg iter 0.006s  
    internal total std dev across ranks 0.000 percent
Fcn/deriv time:     total=0.001s  ( obj=0.000 grad=0.000 cons=0.000 Jac=0.000 Hess=0.000) 
    Fcn/deriv total std dev across ranks 0.000 percent
Fcn/deriv #: obj 16 grad 16 eq cons 16 ineq cons 16 eq Jac 16 ineq Jac 16
Total KKT time 0.079s  
	update init 0.002s     update linsys 0.010s     fact 0.060s  
	solve rhs-manip 0.000s    inner solve 0.007s    resid 0.000s    IR 7.000iters  

selfcheck passed
PASSED: Hiop::test_nlpmdsex1_0
==> [2023-06-13-17:06:04.417047] test: test_nlpmdsex1_1: run example selfcheck
==> [2023-06-13-17:06:04.417690] 'SPACK_ROOT/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/hiop-0.7.2-yjkmatzkgrzjhqjzlw4e62tqpp3xds2t/bin/NlpMdsEx1.exe' '400' '100' '1' '-selfcheck'
===============
Hiop SOLVER
===============
Using 1 MPI ranks.
---------------
Problem Summary
---------------
Total number of variables: 900
     lower/upper/lower_and_upper bounds: 401 / 401 / 1
Total number of equality constraints: 400
Total number of inequality constraints: 3
     lower/upper/lower_and_upper bounds: 2 / 2 / 1
iter    objective     inf_pr     inf_du   lg(mu)  alpha_du   alpha_pr linesrch
   0  3.9800005e+02 4.990e+02  4.000e+00  -1.00  0.000e+00  0.000e+00  -(-)
   1  1.8397701e+01 1.818e+02  1.457e+00  -1.00  3.341e-01  6.357e-01  1(s)
   2 -4.3390041e+01 5.305e+01  4.253e-01  -1.00  7.654e-01  7.081e-01  1(s)
   3 -4.9598500e+01 8.260e-13  2.539e-01  -1.00  8.152e-01  1.000e+00  1(s)
   4 -4.9406538e+01 7.556e-13  3.078e+01  -1.00  9.388e-01  5.909e-02  1(f)
   5 -4.3566407e+01 8.954e-13  1.000e-06  -1.00  1.000e+00  1.000e+00  1(f)
   6 -4.9711482e+01 1.172e-12  1.601e+00  -2.55  9.261e-01  1.000e+00  1(f)
   7 -4.9974904e+01 1.388e-13  2.828e-08  -2.55  1.000e+00  1.000e+00  1(f)
   8 -4.9992701e+01 8.282e-14  1.504e-09  -3.82  1.000e+00  1.000e+00  1(f)
   9 -4.9993494e+01 2.287e-14  1.504e-09  -3.82  1.000e+00  1.000e+00  1(f)
iter    objective     inf_pr     inf_du   lg(mu)  alpha_du   alpha_pr linesrch
  10 -4.9993744e+01 1.221e-14  1.732e-03  -5.73  9.710e-01  1.000e+00  1(f)
  11 -4.9994740e+01 1.342e-12  1.845e-11  -5.73  1.000e+00  1.000e+00  1(f)
  12 -4.9994726e+01 1.354e-14  1.845e-11  -5.73  1.000e+00  1.000e+00  1(f)
  13 -4.9994868e+01 1.110e-14  3.029e-04  -6.00  8.509e-01  1.000e+00  1(f)
  14 -4.9994888e+01 7.421e-13  1.000e-11  -6.00  1.000e+00  1.000e+00  1(f)
Successfull termination.
Total time 0.071s  
Hiop internal time:     total 0.070s      avg iter 0.005s  
    internal total std dev across ranks 0.000 percent
Fcn/deriv time:     total=0.002s  ( obj=0.000 grad=0.000 cons=0.000 Jac=0.001 Hess=0.000) 
    Fcn/deriv total std dev across ranks 0.000 percent
Fcn/deriv #: obj 16 grad 16 eq cons 17 ineq cons 17 eq Jac 16 ineq Jac 16
Total KKT time 0.066s  
	update init 0.001s     update linsys 0.010s     fact 0.049s  
	solve rhs-manip 0.000s    inner solve 0.007s    resid 0.000s    IR 7.000iters  

selfcheck passed
PASSED: Hiop::test_nlpmdsex1_1
==> [2023-06-13-17:06:04.581139] test: test_nlpmdsex1_0-empty_sp_row: run example selfcheck
==> [2023-06-13-17:06:04.581749] 'SPACK_ROOT/opt/spack/linux-rhel8-broadwell/gcc-10.3.1/hiop-0.7.2-yjkmatzkgrzjhqjzlw4e62tqpp3xds2t/bin/NlpMdsEx1.exe' '400' '100' '0' '-empty_sp_row' '-selfcheck'
===============
Hiop SOLVER
===============
Using 1 MPI ranks.
---------------
Problem Summary
---------------
Total number of variables: 900
     lower/upper/lower_and_upper bounds: 401 / 401 / 1
Total number of equality constraints: 400
Total number of inequality constraints: 3
     lower/upper/lower_and_upper bounds: 2 / 2 / 1
iter    objective     inf_pr     inf_du   lg(mu)  alpha_du   alpha_pr linesrch
   0  3.9800005e+02 4.990e+02  4.000e+00  -1.00  0.000e+00  0.000e+00  -(-)
   1 -1.2844361e+01 1.112e+02  4.469e+01  -1.00  1.693e-02  7.772e-01  1(s)
   2 -4.9268731e+01 1.964e-01  2.986e-01  -1.00  8.158e-01  9.982e-01  1(s)
   3 -4.9481416e+01 1.893e-01  2.707e+01  -1.00  7.498e-01  3.622e-02  1(s)
   4 -4.3349942e+01 1.414e-12  1.000e-06  -1.00  1.000e+00  1.000e+00  1(s)
   5 -4.9694295e+01 5.100e-13  1.822e+00  -2.55  9.186e-01  1.000e+00  1(f)
   6 -4.9974131e+01 3.180e-13  2.828e-08  -2.55  1.000e+00  1.000e+00  1(f)
   7 -4.9992656e+01 2.598e-14  1.504e-09  -3.82  1.000e+00  1.000e+00  1(f)
   8 -4.9993484e+01 8.438e-15  1.504e-09  -3.82  1.000e+00  1.000e+00  1(f)
   9 -4.9993742e+01 1.399e-14  1.729e-03  -5.73  9.710e-01  1.000e+00  1(f)
iter    objective     inf_pr     inf_du   lg(mu)  alpha_du   alpha_pr linesrch
  10 -4.9994734e+01 6.861e-13  1.845e-11  -5.73  1.000e+00  1.000e+00  1(f)
  11 -4.9994724e+01 3.131e-14  1.845e-11  -5.73  1.000e+00  1.000e+00  1(f)
  12 -4.9994868e+01 2.154e-14  3.009e-04  -6.00  8.547e-01  1.000e+00  1(f)
  13 -4.9994888e+01 7.851e-13  1.000e-11  -6.00  1.000e+00  1.000e+00  1(f)
Successfull termination.
Total time 0.064s  
Hiop internal time:     total 0.064s      avg iter 0.005s  
    internal total std dev across ranks 0.000 percent
Fcn/deriv time:     total=0.001s  ( obj=0.000 grad=0.000 cons=0.000 Jac=0.000 Hess=0.000) 
    Fcn/deriv total std dev across ranks 0.000 percent
Fcn/deriv #: obj 15 grad 15 eq cons 15 ineq cons 15 eq Jac 15 ineq Jac 15
Total KKT time 0.060s  
	update init 0.001s     update linsys 0.009s     fact 0.044s  
	solve rhs-manip 0.000s    inner solve 0.006s    resid 0.000s    IR 6.500iters  

selfcheck passed
PASSED: Hiop::test_nlpmdsex1_0-empty_sp_row
PASSED: Hiop::test_nlpmdsex1
==> [2023-06-13-17:06:04.735272] test: test_nlpmdsex1raja: run NlpMDsEx1Raja.exe selfchecks
SKIPPED: Hiop::test_nlpmdsex1raja: Tests require +raja
==> [2023-06-13-17:06:04.738152] Completed testing
==> [2023-06-13-17:06:04.738425] 
========================= SUMMARY: hiop-0.7.2-yjkmatz ==========================
Hiop::test_nlpmdsex1_0 .. PASSED
Hiop::test_nlpmdsex1_1 .. PASSED
Hiop::test_nlpmdsex1_0-empty_sp_row .. PASSED
Hiop::test_nlpmdsex1 .. PASSED
Hiop::test_nlpmdsex1raja .. SKIPPED
======================== 4 passed, 1 skipped of 5 parts ========================
============================== 1 passed of 1 spec ==============================
```

### CODE REVIEW

**Comment by cameronrutherford** [comment]
@tldahlgren when will #34236 be merged? I want to add something to exago to support spack test, but am hesitant to do so if it will be refactored again soon. I also wanted to make sure hiop tests were working but these seem like experimental changes.

**Comment by tldahlgren** [comment]
> @tldahlgren when will #34236 be merged? I want to add something to exago to support spack test, but am hesitant to do so if it will be refactored again soon. I also wanted to make sure hiop tests were working but these seem like experimental changes.

Great question.  I'm working on it.  Hopefully within the next week or two.

File: var/spack/repos/builtin/packages/hiop/package.py [hunk 1]
```diff
@@ -232,42 +230,54 @@ def cmake_args(self):
...
     # export SPACK_DISABLE_LOCAL_CONFIG=true
-    def test(self):
-        if not self.spec.satisfies("@develop") or not os.path.isdir(self.prefix.bin):
-            tty.info("Skipping: checks not installed in bin for v{0}".format(self.version))
-            return
+    def test_nlpmdsex1(self):
+        """run NlpMDsEx1.exe selfchecks"""
+        # TODO: These tests should build and run against the installed software.
+        # TODO: Set `test_requires_compiler = True` to pick up them and cmake.
+        # TODO: These tests should *never* run in the installation prefix.
```
Comment (cameronrutherford) [inline var/spack/repos/builtin/packages/hiop/package.py:252]: Are these still TODO?

File: var/spack/repos/builtin/packages/hiop/package.py [hunk 2]
```diff
@@ -232,42 +230,54 @@ def cmake_args(self):
...
         tests = [
-            ["NlpMdsEx1.exe", "400", "100", "0", "-selfcheck"],
-            ["NlpMdsEx1.exe", "400", "100", "1", "-selfcheck"],
-            ["NlpMdsEx1.exe", "400", "100", "0", "-empty_sp_row", "-selfcheck"],
+            ("NlpMdsEx1.exe", ["400", "100", "0", "-selfcheck"]),
+            ("NlpMdsEx1.exe", ["400", "100", "1", "-selfcheck"]),
+            ("NlpMdsEx1.exe", ["400", "100", "0", "-empty_sp_row", "-selfcheck"]),
         ]
+        for exe, args in tests:
+            with test_part(
```
Comment (cameronrutherford) [inline var/spack/repos/builtin/packages/hiop/package.py:260]: I am unfamiliar with `test_part` here.

**Review by cameronrutherford — APPROVED** [review]
A few points of clarification, but LGTM! We should surely apply similar tests to ExaGO package.py in the future

File: var/spack/repos/builtin/packages/hiop/package.py (hunk 2 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/hiop/package.py:260]: It's the new API for embedding test (sub)parts in an implicit test part (or `test_*` method).  The docs describe it in the section at  https://spack.readthedocs.io/en/latest/packaging_guide.html#adding-stand-alone-tests.

File: var/spack/repos/builtin/packages/hiop/package.py (hunk 1 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/hiop/package.py:252]: That's a great question.  Is the example built against the library and headers under `self.prefix`?  Or is it built then relocated to `self.prefix.bin`?

Comment (cameronrutherford) [reply var/spack/repos/builtin/packages/hiop/package.py:252]: Apologies for the delay. These tests are built against the shared/static library that is present in the `install/lib` directory. I assume this is `self.prefix.lib`? The executable is installed into the `install/bin` directory, which I assume is `self.prefix.bin` (https://github.com/LLNL/hiop/blob/4b231f71dd7aa7994d841b2fa0b52ea60317f6d0/src/Drivers/MDS/CMakeLists.txt#L3).

File: var/spack/repos/builtin/packages/hiop/package.py (hunk 2 above)
Comment (cameronrutherford) [reply var/spack/repos/builtin/packages/hiop/package.py:260]: Ok I will probably have to re-visit this when adding tests to ExaGO, but I really appreciate this :)

File: var/spack/repos/builtin/packages/hiop/package.py (hunk 1 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/hiop/package.py:252]: The package's `self.prefix` is, by default, in a subdirectory under `$SPACK_ROOT/opt/spack/architecture>/<compiler>/<package-version-hash>` but can be set in the configuration with `config:install_tree:root` (see https://spack.readthedocs.io/en/latest/config_yaml.html#config-yaml and https://spack.readthedocs.io/en/latest/config_yaml.html#install-tree-root).  AFAICT. The example is built at install time.  

What I am suggesting is to have a program that can be built and run at stand-alone test time (so days to weeks after the installation).

There's a very simple example to illustrate in the docs at https://spack.readthedocs.io/en/latest/packaging_guide.html#enabling-test-compilation.

File: var/spack/repos/builtin/packages/hiop/package.py [hunk 3]
```diff
@@ -232,42 +230,54 @@ def cmake_args(self):
 
         return args
 
+    def _run_test_example(self, test_exe, options):
+        """Run the provided test executable with the provided options."""
+
+        exe_path = os.path.join(self.prefix.bin, test_exe)
+        if not os.path.exists(exe_path):
+            raise SkipTest("{0} is not installed".format(test_exe))
```
Comment (cameronrutherford) [inline var/spack/repos/builtin/packages/hiop/package.py:238]: Finally understanding this more... I imagine that we could embed the logic into each test to only run if the right set of variants are configured for hiop, but this is less error prone :)

File: var/spack/repos/builtin/packages/hiop/package.py [hunk 4]
```diff
@@ -232,42 +230,54 @@ def cmake_args(self):
...
+                self,
+                "test_nlpmdsex1_{0}".format("".join(args[2:-1])),
+                purpose="run example selfcheck",
+            ):
+                self._run_test_example(exe, args)
+
+    def test_nlpmdsex1raja(self):
+        """run NlpMDsEx1Raja.exe selfchecks"""
+        if "+raja" not in self.spec:
+            raise SkipTest("Tests require +raja")
```
Comment (cameronrutherford) [inline var/spack/repos/builtin/packages/hiop/package.py:270]: I suppose this is also easy to add in simple cases like this... Either way the same SkipTest exception is raised, so it is nice to have flexibility

File: var/spack/repos/builtin/packages/hiop/package.py (hunk 1 above)
Comment (cameronrutherford) [reply var/spack/repos/builtin/packages/hiop/package.py:252]: I think we only need to worry about this when building shared objects? I am still developing my understanding here.

I am guessing that static objects have everything that they need to run, whereas shared objects will need to have the CXX compiler available on the path?

**Review by cameronrutherford — COMMENTED** [review]
I appreciate the help here @tldahlgren

**Comment by tldahlgren** [comment]
Replaced by #44765

**Pull request closed without merging**